src/main/java/org/folio/circulation/services/FeeFineService.java

```diff
@@ -1,58 +1,164 @@
 package org.folio.circulation.services;
 
 import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.services.feefine.AccountRefundProcessors.getProcessor;
+import static org.folio.circulation.support.Result.combined;
+import static org.folio.circulation.support.Result.failed;
+import static org.folio.circulation.support.Result.succeeded;
 
 import java.util.Collection;
+import java.util.List;
 import java.util.concurrent.CompletableFuture;
 
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.folio.circulation.domain.Account;
 import org.folio.circulation.domain.AccountRepository;
-import org.folio.circulation.domain.FeeFineAccountAndAction;
 import org.folio.circulation.domain.FeeFineActionRepository;
-import org.folio.circulation.domain.representations.AccountStorageRepresentation;
-import org.folio.circulation.domain.representations.FeeFineActionStorageRepresentation;
+import org.folio.circulation.domain.ServicePoint;
+import org.folio.circulation.domain.ServicePointRepository;
+import org.folio.circulation.domain.User;
+import org.folio.circulation.domain.UserRepository;
+import org.folio.circulation.domain.representations.StoredAccount;
+import org.folio.circulation.domain.representations.StoredFeeFineAction;
+import org.folio.circulation.services.feefine.AccountRefundContext;
+import org.folio.circulation.services.feefine.AccountRefundProcessor;
+import org.folio.circulation.services.support.CreateAccountCommand;
+import org.folio.circulation.services.support.RefundAccountCommand;
 import org.folio.circulation.support.Clients;
 import org.folio.circulation.support.Result;
 import org.folio.circulation.support.results.CommonFailures;
 
+import io.vertx.core.json.JsonObject;
+
 public class FeeFineService {
   private final AccountRepository accountRepository;
   private final FeeFineActionRepository feeFineActionRepository;
+  private final UserRepository userRepository;
+  private final ServicePointRepository servicePointRepository;
 
   public FeeFineService(Clients clients) {
     this.accountRepository = new AccountRepository(clients);
     this.feeFineActionRepository = new FeeFineActionRepository(clients);
+    this.userRepository = new UserRepository(clients);
+    this.servicePointRepository = new ServicePointRepository(clients);
   }
 
   public CompletableFuture<Result<Void>> createAccountsAndActions(
-    Collection<FeeFineAccountAndAction> accountAndActions) {
+    Collection<CreateAccountCommand> accountAndActions) {
 
     return allOf(accountAndActions.stream()
       .map(this::createAccountAndAction)
       .toArray(CompletableFuture[]::new))
-      .thenApply(notUsed -> Result.<Void>succeeded(null))
+      .thenApply(Result::succeeded)
       .exceptionally(CommonFailures::failedDueToServerError);
```

Comment:
Why are these steps needed when the `createAccountAndAction` method returns a `CompletableFuture<Result<Void>>` already? Is this to convert a stream of results to a single result for all of them?

Reply:
Yes, correct. The method returns single completable future which is successful only when all of the futures are completed successfully.

Reply:
What happens during this process when one of the commands fails? From what I can tell, this will create a successful result, masking that failure.

Reply:
`thenApply(Result::succeeded)` won't be executed when a command failed, but the `exceptionally(CommonFailures::failedDueToServerError)` method will be executed instead.

Reply:
Does that mean that `createAccount` will throw an exception when it fails, rather than return a failed result?

As I understand it, the `thenApply` will only be skipped and `exceptionally` executed when an unhandled exception is thrown. Have I misunderstood?

Reply:
Thanks for pointing on this. It was not working as I expected to. I've added unit test for both methods and fixed methods.